yarn-project/archiver/src/archiver/archiver.ts

```diff
@@ -1,18 +1,46 @@
-import { AztecAddress, EthAddress, createDebugLogger } from '@aztec/foundation';
+import { AztecAddress, EthAddress, RunningPromise, createDebugLogger } from '@aztec/foundation';
+import { INITIAL_L2_BLOCK_NUM } from '@aztec/l1-contracts';
 import { RollupAbi, UnverifiedDataEmitterAbi } from '@aztec/l1-contracts/viem';
-import { ContractData, L2Block, L2BlockSource } from '@aztec/types';
-import { createPublicClient, decodeFunctionData, getAddress, Hex, hexToBytes, http, Log, PublicClient } from 'viem';
+import { ContractData, L2Block, L2BlockSource, UnverifiedData, UnverifiedDataSource } from '@aztec/types';
+import {
+  Chain,
+  Hex,
+  HttpTransport,
+  Log,
+  PublicClient,
+  createPublicClient,
+  decodeFunctionData,
+  getAbiItem,
+  getAddress,
+  hexToBytes,
+  http,
+} from 'viem';
 import { localhost } from 'viem/chains';
 import { ArchiverConfig } from './config.js';
-import { INITIAL_L2_BLOCK_NUM } from '@aztec/l1-contracts';
-import { UnverifiedData, UnverifiedDataSource } from '@aztec/types';
 
 /**
  * Pulls L2 blocks in a non-blocking manner and provides interface for their retrieval.
- * Responsible for handling robust L1 polling (TODO) so that other components do not
- * need to concern themselves with it.
+ * Responsible for handling robust L1 polling so that other components do not need to
+ * concern themselves with it.
  */
 export class Archiver implements L2BlockSource, UnverifiedDataSource {
+  /**
+   * The number of blocks that we will fetch in a single batch.
+   */
+  private readonly numL1BlocksPerFetch = 1000n;
+
+  /**
+   * A range of blocks in which we expect a potential node sync issue as described in
+   * https://github.com/wagmi-dev/viem/discussions/354
+   * Danger zone is defined as [BLOCK_HEIGHT - dangeZoneBlockLength, BLOCK_HEIGHT].
+   */
+  private readonly dangerZoneBlockLength = 1000n;
```

Comment:
I'm not sure we need this. If the last rollup was published at eth block n and we make sure we always scan from eth block n + 1 then I think that is sufficient. Node syncing problems don't mean we have to 'avoid' scanning a range of recent blocks, it just means we can't rely on a negative log response being correct for recent blocks. As long as we always rescan from the last known rollup block then we will be covered. The 'missing' log will turn appear in subsequent re-scans.

Reply:
I saw in the docs that some nodes don't allow you to scan more than a certain amount of blocks. This led me to introducing this distinction between initial sync finishing at the beginning of the  "danger zone" and normal sync without danger zone. In the initial sync I rely on negative log response which should be ok given that it's far from the latest block height.

Reply:
Yes I think we can simplify it. Whether it is the initial sync or otherwise, we should be fine to simply scan up to the current block.